### ISSUE TITLE
Volume adjustments queue until screen brightens in Lock Screen

### ISSUE THREAD

**Issue by fivecar** [doublesymmetry/react-native-track-player#2079]
**Describe the Bug**
In v4rc6, if you `setVolume` while your phone is locked and the screen is off, the volume change will only take effect once your screen comes back on, even though playback proceeds just fine.

**Steps To Reproduce**
Run in production mode, detached from Xcode. Basically:

1. Write some code that, when the headset presses play, ramps the volume up from 0.1 to 1.0 over two seconds.
2. Start your app, then put your phone into the lock screen and wait for the screen to turn off.
3. Press Play on your headset

RESULT:
Playback stays at 0.1 volume. Oddly, it ramps into 1.0 if you turn the screen on (e.g. by pressing the lock button on your phone, or picking your phone up so that the screen turns on). Until the screen brightens, the volume stays at 0.1. But once the screen brightens, even if the phone is still locked, the volume ramps to 1.0 as expected.

**Note**: This was not a bug back when I ran 3.2.0 on iOS 16. However, I'm now running v4 rc6 on iOS 17 beta, and it happens.

**Code To Reproduce**
This is the code which I've run that reproduces this (in production only -- doesn't reproduce in debug builds while connected to Xcode):

```typescript
function onRemotePlay() {
      const PLAY_FADE_STEPS = 10;
      const volumes = Array(PLAY_FADE_STEPS)
        .fill(0)
        .map((_, i) => (i + 1) / PLAY_FADE_STEPS);

      // RxJS -- I hope this makes sense to everybody. But basically this fades the volume in from 0.1 to 1.0 over 2 secs.
      from(volumes)
        .pipe(
          concatMap(async v => {
            await TrackPlayer.setVolume(v);  // Sets volume from 0.1, 0.2, 0.3... to 1.0
            await sleep(2000 / PLAY_FADE_STEPS);
            return v;
          }),
        )
        .subscribe({
          error: err => {
            logError(ErrorEvent.ASSERT_FAILURE, {
              error: `(play) Fade-in: ${errToString(err)}`,
            });

            // Always restore volume, no matter what.
            void TrackPlayer.setVolume(1);
          },
        });
}
```

**Replicable on Example App?**
When I try this in the example app, it doesn't reproduce. I'm not sure why this is the case. But here's the code that I use in the example app:
```typescript
async function onRemotePlay() {
  console.log('Remote play received');

  const PLAY_FADE_STEPS = 10;
  const volumes = Array(PLAY_FADE_STEPS)
    .fill(0)
    .map((_, i) => (i + 1) / PLAY_FADE_STEPS);

  await TrackPlayer.setVolume(0.1);
  await TrackPlayer.play();

  for (const volume of volumes) {
    await new Promise<void>((resolve) => setTimeout(resolve, 500));
    console.log('Attempting to set volume', volume);
    await TrackPlayer.setVolume(volume);
  }
}
```

**Environment Info:**
Real device, running iOS 17 Beta. (Not sure if this might have to do with it). RNTP RC6.
```
System:
    OS: macOS 13.5
    CPU: (12) x64 Apple M2 Max
    Memory: 757.43 MB / 64.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 16.16.0 - ~/.nvm/versions/node/v16.16.0/bin/node
    Yarn: 3.5.0 - ~/.nvm/versions/node/v16.16.0/bin/yarn
    npm: 9.5.0 - ~/.nvm/versions/node/v16.16.0/bin/npm
    Watchman: Not Found
  Managers:
    CocoaPods: 1.11.3 - /Users/philipsu/.rvm/gems/ruby-2.7.6/bin/pod
  SDKs:
    iOS SDK:
      Platforms: DriverKit 22.4, iOS 16.4, macOS 13.3, tvOS 16.4, watchOS 9.4
    Android SDK: Not Found
  IDEs:
    Android Studio: 2020.3 AI-203.7717.56.2031.7784292
    Xcode: 14.3.1/14E300c - /usr/bin/xcodebuild
  Languages:
    Java: 17.0.4.1 - /usr/bin/javac
  npmPackages:
    @react-native-community/cli: Not Found
    react: 18.2.0 => 18.2.0 
    react-native: 0.71.8 => 0.71.8 
    react-native-macos: Not Found
```

**How I can Help**
I'd love to help on this, and can submit a PR if someone points me in the right direction. At the moment, I don't have a mental model about what could be going on (e.g. what possible difference is turning on the screen vs. having the screen off?). It's weird that the volume ramps immediately to 1.0 the moment the phone's screen comes on. But I don't understand what could make the volume suddenly ramp to 1.0 if the screen comes on, yet stay at volume 0.1 as long as the screen remains black.

Thanks for any help/input.

**Comment by fivecar** [comment]
Could this somehow be because iOS 17 doesn't execute `setTimeout` handlers when the screen is off, but does execute your handlers on the lock screen if the screen is on? That'd be kind of a weird policy, IMO, but it'd explain the behavior I'm seeing (though it still doesn't explain why it doesn't repro in the example app).

**Comment by fivecar** [comment]
@dcvz - hey, really sorry, had started a new job and lost this thread for a while.

This repros for me 100%, and unfortunately hits the [production version of my app](https://apps.apple.com/us/app/superphonic/id6444668430) consistently, all the time.

It did not happen for me on iOS 16, but was consistent in iOS 17 through all the betas, and now in the officially launched iOS 17.

Do you have any theories of what I should potentially look for in order to debug this issue? I know my side of things pretty well (the React Native app), but I don't have a good mental model of what this library, along with SwiftAudioEx, do behind the scenes. Is there anything in those libraries that could be impacted by iOS 17 when it comes to volume?

On my side, the code that calls `TrackPlayer.play()` is in the same block of code that ramps the volume up via `TrackPlayer.setVolume()`, so the background event handler is definitely running successfully. I don't think iOS 17 would queue `setTimeout` tasks any differently depending on whether your app is backgrounded, since that'd break a lot more apps. So it leaves me thinking that it might be something about AVPlayer or SwiftAudioEx's calling of it. Not to mention the oddity that the volume ramps up immediately the moment you get the screen to light up (e.g. by moving/tilting the phone or by pressing the power button) -- which really does not feel like something that React Native would control.

**Comment by fivecar** [comment]
@dcvz I'm going to update to rc8 and report back. FWIW, rc6 has been pretty solid for me, with the exception of a known issue (`AVPlayerItem addOutput` crash). 🤞

**Comment by fivecar** [comment]
Ooh, hey -- new theory: I wonder if iOS 17 is more aggressively backgrounding apps and not running their `setTimeout` calls. Perhaps when the screen lights up, the OS begins calling `setTimeout` callbacks again.

To test this theory, I'm integrating [react-native-background-timer](https://github.com/ocetnik/react-native-background-timer) to see if it helps. I'll report back after testing.

**Comment by fivecar** [comment]
@dcvz Ok - I've run it a few days with a background timer, and things appear to be working just fine! I'm chalking this up to user error. No idea why: a) this didn't reproduce in iOS 16, and b) this didn't reproduce in the example app (because it really should have, if the operant theory is that `setTimeout` isn't getting called until the phone lights up).

**Comment by Ggayane** [comment]
I've recently updated the package from `3.2.0-9cd6e0e9235f8ed7ae5812ed20220f7d895b510a` to `v4` and noticed on both android and ios  when app is on the background, background activities just stop working, once I open app everything start to load and work fine again. Does anyone faced with is?

An issue described  above looked similar to what I'm facing... But I have no idea what has been changed, couldn't find anything in release notes neither.
